update Scanner with additional inputs

cbeer opened the pull request:
expands Scanner signature to support:
- an IO (e.g. File, STDIN, etc)
- a path to Gemfile.lock
- a path to directory

Or any array of those.

cbeer committed:
update Scanner with additional inputs

expands Scanner signature to support:
 - an IO (e.g. File, STDIN, etc)
 - a path to Gemfile.lock
 - a path to directory

Or any array of those.

mackuba commented on lib/holepicker/scanner.rb:
```diff
@@ -14,8 +14,11 @@ class Scanner
+
+      if options[:stdin]
+        @paths = STDIN
+      end
```
I think you can just change the `bot` script to pass `STDIN` as the argument instead of a `:stdin => true` option and simplify these few lines here, I don't think we need to treat that option as a public API...

cbeer replied:
Sounds good, will do.

mackuba commented on lib/holepicker/scanner.rb:
```diff
@@ -50,15 +49,31 @@ def scan
 
+    def paths
+      @paths.is_a?(Array) ? @paths : [@paths]
+    end
```
Any reason why this would be better than assigning this once in the constructor? `paths` is not a public property so it won't be modified.

cbeer replied:
I'll take a look, and add a comment if I remember why.

cbeer replied:
(It was like this to simplify the initializer logic in the `options[:stdin]` case. No special case, no problem.)

mackuba commented on lib/holepicker/scanner.rb:
```diff
@@ -50,15 +49,31 @@ def scan
       @database.vulnerabilities.select { |v| v.gem_vulnerable?(gem) }
     end
 
+    def scan_something p
```
Maybe `scan_source(source)`? "something" doesn't sound right :) (also, most style guides discourage such definitions without parentheses around arguments)

cbeer replied:
Works for me, will do.

mackuba commented on lib/holepicker/scanner.rb:
```diff
@@ -50,15 +49,31 @@ def scan
+        else
+          scan_path(p)
+        end
+    end
```
I think `when` (and `else`, `end`) directly below `case` (i.e. not indented) is a more official style (e.g. https://github.com/bbatsov/ruby-style-guide). Sorry, I'm a bit OCD about formatting :)

mackuba commented on lib/holepicker/scanner.rb:
```diff
@@ -50,15 +49,31 @@ def scan
+
     def scan_path(path)
-      @finder.find_gemfiles(path).each { |f| scan_gemfile(File.read(f), f) }
+        @finder.find_gemfiles(path).each { |f| scan_gemfile(File.read(f), f) }
```
And this looks like an accident...

mackuba commented:
Good idea in general, thanks for sending the patch.

cbeer commented:
I've fixed the issues noted and updated the pull request. If it's valuable, I wrote a couple integration tests against the `bin/holepicker` utility when making the `STDIN` change, but it is very slow to run (1.5s locally). I'm happy to commit it and make it only run under travis, or some other condition, if you're interested.

Alternatively, here's the result of manual testing:

```
~/Projects/holepicker $ bundle exec bin/holepicker Gemfile.lock 
Fetching list of vulnerabilities...
Gemfile.lock: ✔
No vulnerabilities found.

~/Projects/holepicker $ bundle exec bin/holepicker .
Fetching list of vulnerabilities...
/Volumes/TempStorage/Projects/holepicker/Gemfile.lock: ✔
No vulnerabilities found.

~/Projects/holepicker $ bundle exec bin/holepicker --stdin < Gemfile.lock 
Fetching list of vulnerabilities...
Scanning gemfile: ✔
No vulnerabilities found.
```
